stdlib/public/core/ArrayBufferProtocol.swift

```diff
@@ -154,66 +154,37 @@ extension _ArrayBufferProtocol {
       self.count = oldCount + growth
     }
 
-    let elements = self.subscriptBaseAddress
-    let oldTailIndex = subrange.upperBound
-    let oldTailStart = elements + oldTailIndex
-    let newTailIndex = oldTailIndex + growth
-    let newTailStart = oldTailStart + growth
-    let tailCount = oldCount - subrange.upperBound
-
-    if growth > 0 {
-      // Slide the tail part of the buffer forwards, in reverse order
-      // so as not to self-clobber.
-      newTailStart.moveInitialize(from: oldTailStart, count: tailCount)
-
-      // Update the original subrange
-      var i = newValues.startIndex
-      for j in subrange {
-        elements[j] = newValues[i]
-        newValues.formIndex(after: &i)
-      }
-      // Initialize the hole left by sliding the tail forward
-      for j in oldTailIndex..<newTailIndex {
-        (elements + j).initialize(to: newValues[i])
-        newValues.formIndex(after: &i)
-      }
-      _expectEnd(of: newValues, is: i)
-    }
-    else { // We're not growing the buffer
-      // Assign all the new elements into the start of the subrange
-      var i = subrange.lowerBound
-      var j = newValues.startIndex
-      for _ in 0..<newCount {
-        elements[i] = newValues[j]
-        i += 1
-        newValues.formIndex(after: &j)
-      }
-      _expectEnd(of: newValues, is: j)
+    let elements = self.firstElementAddress
 
-      // If the size didn't change, we're done.
-      if growth == 0 {
-        return
-      }
+    // erase all the elements we're replacing to create a hole
+    let holeStart = elements + subrange.lowerBound
+    let holeEnd = holeStart + newCount
 
-      // Move the tail backward to cover the shrinkage.
-      let shrinkage = -growth
-      if tailCount > shrinkage {   // If the tail length exceeds the shrinkage
+    if eraseCount > 0 {
+      holeStart.deinitialize(count: eraseCount)
+    }
 
-        // Update the rest of the replaced range with the first
-        // part of the tail.
-        newTailStart.moveUpdate(from: oldTailStart, count: shrinkage)
+    // resize the hole to make it the correct size
+    if growth != 0 {
+      let tailStart = elements + subrange.upperBound
+      let tailCount = oldCount - subrange.upperBound
+      holeEnd.moveInitialize(from: tailStart, count: tailCount)
+    }
 
-        // Slide the rest of the tail back
-        oldTailStart.moveInitialize(
-          from: oldTailStart + shrinkage, count: tailCount - shrinkage)
+    // place the values into the hole we created
+    if newCount > 0 {
+      let done: Void? = newValues.withContiguousStorageIfAvailable {
+        holeStart.initialize(from: $0.baseAddress!, count: newCount)
```

Comment:
It's so annoying that withContiguousStorageIfAvailable is allowed to pass a buffer with a nil baseAddress rather than just returning nil without executing the closure in that case. Otherwise we could use .unsafelyUnwrappedUnchecked here.

Reply:
(Actually what IS supposed to happen in that case? Should it just be the same as if newValues was empty?)

Reply:
well, it says that if there is no contiguous representation, it _should_ ignore body and just return nil?
_and_ in this case, `newValues.count` / `newCount` _should_ be the same for a well-implemented collection?

I'm not sure if its worth throwing in an _extra_ assert for $0.count == newCount, or removing `!` for `.unsafelyUnwrappedUnchecked` either.